tlscanary/modes/regression.py

```diff
@@ -130,99 +133,82 @@ def run(self):
-                                       prefs=self.args.prefs_base, report_callback=report_overhead)
-        logger.info("Second baseline candidate pass yielded %d error hosts" % len(base_error_set))
-        logger.debug("Second baseline candidate pass errors: %s"
-                     % ' '.join(["%d,%s" % (r, u) for r, u in base_error_set]))
-
-        error_set = test_error_set.difference(base_error_set)
-
-        # Third pass:
-        # - Run error set from first pass against the test candidate with less workers
-        # - Run new error set against baseline candidate with less workers
-        # - Filter for errors from test candidate but not baseline
-
-        logger.info("Starting third pass with %d hosts" % len(error_set))
-
-        test_error_set = self.run_test(self.test_app, error_set, profile=self.test_profile,
-                                       prefs=self.args.prefs_test, num_workers=2, n_per_worker=10,
-                                       report_callback=report_overhead)
-        logger.info("Third test candidate pass yielded %d error hosts" % len(test_error_set))
-        logger.debug("Third test candidate pass errors: %s"
-                     % ' '.join(["%d,%s" % (r, u) for r, u in test_error_set]))
-
-        base_error_set = self.run_test(self.base_app, test_error_set, profile=self.base_profile,
-                                       prefs=self.args.prefs_base, num_workers=2, n_per_worker=10,
-                                       report_callback=report_overhead)
-        logger.info("Third baseline candidate pass yielded %d error hosts" % len(base_error_set))
-        logger.debug("Third baseline candidate pass errors: %s"
-                     % ' '.join(["%d,%s" % (r, u) for r, u in base_error_set]))
-
-        error_set = test_error_set.difference(base_error_set)
-        if len(error_set) > 0:
-            logger.warning("%d regressions found: %s"
-                           % (len(error_set), ' '.join(["%d,%s" % (r, u) for r, u in error_set])))
-
-        # Fourth pass, information extraction:
-        # - Run error set from third pass against the test candidate with less workers
+        # Compile set of error hosts in multiple scans
+
+        # Each scan:
+        # - Runs full test set against the test candidate
+        # - Runs new error set against baseline candidate
+        # - Take any remaining errors and repeat the above steps
+
+        current_host_set = host_set
+        current_scan = 0
+
+        while current_scan < self.args.scans:
+            # Slow down number of workers and scans with each pass
+            # to make results more precise
+            scans_remaining = self.args.scans - current_scan
+            current_scan += 1
+            pct_remaining = (float(scans_remaining) / float(self.args.scans))
+            adjusted_workers = int(ceil(self.args.parallel * pct_remaining))
+            adjusted_requests_per_worker = int(ceil(self.args.requestsperworker * pct_remaining))
```

Comment:
The way for calculating a round-based number of workers / and or the timeout seems rather complex. I'd use an exponential instead of a linear approach like so:

```python
num_workers = self.args.parallel
requests_per_worker = self.args.requestsperworker
timeout = self.args.timeout

for current_scan in xrange(1, self.args.scans+1):
    [...]
    "do your things"
    [...]
    num_workers = max(1, int(num_workers * 0.75))
    requests_per_worker = max(1, int(requests_per_worker * 0.75))
    timeout = min(SOME_MAXIMUM_TIMEOUT, timeout * 1.5))
```

The exponents were chosen more or less randomly and would likely require adjustment. Unless there are good reasons to choose a linear scaling, this approach would IMHO add significantly less complexity.